server/src/main/java/org/opensearch/index/shard/StoreRecovery.java

```diff
@@ -268,7 +268,9 @@ public void copyFrom(Directory from, String src, String dest, IOContext context)
             in.copyFrom(new FilterDirectory(from) {
                 @Override
                 public IndexInput openInput(String name, IOContext context) throws IOException {
-                    index.addFileDetail(dest, l, false);
+                    if (index.getFileDetails(dest) == null) {
```

Comment:
Just a question - this check has been added to dedup file download? or from stats point of view?

Reply:
Why do we need to add this ? Is it for retries for the same file ?

Reply:
This check is from stats point of view. If I report the same file twice, [assertion](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/replication/common/ReplicationLuceneIndex.java#L352) in `addFileDetails` method fails.

Before downloading files, we are reporting all the files to be downloaded (new code block is added in the `IndexShard` class) before the actual transfer starts.

Reply:
Should we then remove the assert ? Looks to be harmless .

Reply:
Those asserts make sure that we don't add the same file twice to the recovery tracker. IMO, that is the correct behavior. Currently, in our restore flow, we call `IndexShard.syncSegmentsFromRemoteSegmentStore` twice. Even though the second time, we skip downloading the files, without the above check, it will get added to tracker again.